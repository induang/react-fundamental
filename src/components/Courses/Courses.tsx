import { useState, useEffect } from 'react';
import { Link } from 'react-router-dom';
import { Box } from '@mui/system';
import { Button, Grid } from '@mui/material';
import { useAppDispatch, useAppSelector } from '../../redux/store/storeHook';

import SearchBar from './components/SearchBar/SearchBar';
import CourseCard from './components/CourseCard/CourseCard';
import { AppDispatch, RootState } from '../../types';
import { Course, getCoursesThunk } from '../../redux/slices/courseSlice';
import { authorizeThunk } from '../../redux/slices/userSlice';

const Courses: React.FC = () => {
	const courses = useAppSelector((state: RootState) => state.courses.courses);
	const dispatch: AppDispatch = useAppDispatch();
	const [curCourses, setCurCourses] = useState(courses);

	useEffect(() => {
		dispatch(getCoursesThunk());
		dispatch(authorizeThunk());
	}, []);

	useEffect(() => {
		setCurCourses(courses);
	}, [courses]);

	function courseFilter(course: Course, value: string) {
		return (
			course.title.toLowerCase().search(new RegExp(value, 'i')) !== -1 ||
			course.id.search(value) !== -1
		);
	}

	function handleSearch(value: string) {
		setCurCourses([
			...courses.filter((course: Course) => courseFilter(course, value)),
		]);
	}

	return (
		<Box padding={3}>
			<Box>
				<Box>
					<Grid container className='flex justify-between'>
						<Grid item>
							<SearchBar handleSearch={handleSearch} />
						</Grid>
						<Grid item>
							<Link to='add'>
								<Button variant='outlined' color='secondary'>
									Add new course
								</Button>
							</Link>
						</Grid>
					</Grid>
				</Box>
				{/* 课程列表渲染 */}
				<Box>
					{curCourses && curCourses.length ? (
						curCourses.map((course: Course) => (
							<CourseCard key={course.id} course={course} />
						))
					) : (
						<span>Loading...</span>
					)}
				</Box>
			</Box>
		</Box>
	);
};
export default Courses;
