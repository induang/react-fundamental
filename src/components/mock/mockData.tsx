export const mockedCourse = {
	id: 'testID',
	title: 'testTitle',
	description: 'testDescrip',
	creationDate: '09/10/2022',
	duration: 12,
	authors: ['testAuthorId_1', 'testAuthorId_2'],
};

export const mockedAuthor_1 = {
	id: 'testAuthorId_1',
	name: 'testAuthorName_1',
};

export const mockedAuthor_2 = {
	id: 'testAuthorId_2',
	name: 'testAuthorName_2',
};

export const mockedUser = {
	isAuth: true,
	name: 'TestName',
	email: 'test',
	token: 'test',
	role: 'test',
};
