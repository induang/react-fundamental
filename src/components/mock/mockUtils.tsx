import { PropsWithChildren } from 'react';
import { Provider } from 'react-redux';
import { configureStore, PreloadedState } from '@reduxjs/toolkit';
import { render, RenderOptions } from '@testing-library/react';
import { ToolkitStore } from '@reduxjs/toolkit/dist/configureStore';

import { RootState } from '../../redux/store/store';
import userSlice from '../../redux/slices/userSlice';
import courseSlice from '../../redux/slices/courseSlice';
import authorsSlice from '../../redux/slices/authorsSlice';

import {
	mockedUser,
	mockedCourse,
	mockedAuthor_1,
	mockedAuthor_2,
} from './mockData';

interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
	preloadedState?: PreloadedState<RootState>;
	store?: ToolkitStore;
}
export function renderWithProviders(
	ui: React.ReactElement,
	{
		preloadedState = {
			user: mockedUser,
			courses: {
				courses: [mockedCourse],
			},
			authors: {
				authors: [mockedAuthor_1, mockedAuthor_2],
			},
		},
		store = configureStore({
			reducer: { user: userSlice, courses: courseSlice, authors: authorsSlice },
			preloadedState,
		}),
		...renderOptions
	}: ExtendedRenderOptions = {}
) {
	function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
		return <Provider store={store}>{children}</Provider>;
	}
	return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}
