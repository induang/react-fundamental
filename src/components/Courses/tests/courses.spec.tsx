import { BrowserRouter as Router } from 'react-router-dom';

import { renderWithProviders } from '../../mock/mockUtils';
import {
	mockedAuthor_1,
	mockedAuthor_2,
	mockedCourse,
} from '../../mock/mockData';
import Courses from '../Courses';
import userEvent from '@testing-library/user-event';
import { waitFor } from '@testing-library/react';

window.alert = jest.fn(() => ({}));
jest.mock('../../../services', () => {
	return {
		async fetchAuthorsData() {
			return {
				result: [mockedAuthor_1, mockedAuthor_2],
			};
		},
		async fetchCoursesData() {
			return {
				result: [mockedCourse],
			};
		},
		async fetchAuthorization() {
			return {
				result: {
					isAuth: true,
					name: 'testName',
					email: 'testName@email.com',
					role: 'admin',
				},
			};
		},
	};
});

describe('render courses', () => {
	it('render correctly', () => {
		const { container } = renderWithProviders(
			<Router>
				<Courses />
			</Router>
		);
		expect(container.firstChild).toMatchSnapshot();
	});

	it('display amount of CourseCard equal length of courses array', () => {
		const { store, queryAllByRole } = renderWithProviders(
			<Router>
				<Courses />
			</Router>
		);
		expect(queryAllByRole('heading', { level: 1 })).toHaveLength(
			store.getState().courses.courses.length
		);
	});
	it('shown after a click on the "Add new Course" button', async () => {
		expect.assertions(6);
		const { getByRole, findByRole } = renderWithProviders(
			<Router>
				<Courses />
			</Router>
		);
		const button = getByRole('button', {
			name: /add new course/i,
		});
		userEvent.click(button);

		await waitFor(() => {
			expect(findByRole('heading', { name: /title/i })).toBeTruthy();
			expect(findByRole('heading', { name: /description/i })).toBeTruthy();
			expect(findByRole('heading', { name: /add author/i })).toBeTruthy();
			expect(findByRole('heading', { name: /duration/i })).toBeTruthy();
			expect(findByRole('heading', { name: /authors/i })).toBeTruthy();
			expect(findByRole('heading', { name: /course authors/i })).toBeTruthy();
		});
	});
});
