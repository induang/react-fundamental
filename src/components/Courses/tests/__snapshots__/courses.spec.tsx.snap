// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render courses render correctly 1`] = `
<div
  class="MuiBox-root css-19kzrtu"
>
  <div
    class="MuiBox-root css-0"
  >
    <div
      class="MuiBox-root css-0"
    >
      <div
        class="MuiGrid-root MuiGrid-container flex justify-between css-11lq3yg-MuiGrid-root"
      >
        <div
          class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
        >
          <form
            class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 flex items-center w-96 py-0.5 px-1 css-1ps6pg7-MuiPaper-root"
          >
            <div
              class="MuiInputBase-root MuiInputBase-colorPrimary w-full ml-2 flex css-1le28he-MuiInputBase-root"
            >
              <input
                class="MuiInputBase-input css-yz9k0d-MuiInputBase-input"
                placeholder="Enter course name..."
                type="text"
                value=""
              />
            </div>
            <button
              aria-label="search"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium p-2.5 css-78trlr-MuiButtonBase-root-MuiIconButton-root"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="SearchIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
                />
              </svg>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </form>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
        >
          <a
            href="/add"
          >
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedSecondary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-root MuiButton-outlined MuiButton-outlinedSecondary MuiButton-sizeMedium MuiButton-outlinedSizeMedium css-wnpvqa-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="button"
            >
              Add new course
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </a>
        </div>
      </div>
    </div>
    <div
      class="MuiBox-root css-0"
    >
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation2 my-1 p-1 css-1mbunh-MuiPaper-root"
      >
        <div
          class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 p-5 css-mhc70k-MuiGrid-root"
        >
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-md-7 css-3vt552-MuiGrid-root"
          >
            <h1
              class="text-4xl font-black"
            >
              testTitle
            </h1>
            <div
              class="w-11/12 indent-2"
            >
              testDescrip
            </div>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-md-5 flex-col justify-center css-yw6hac-MuiGrid-root"
          >
            <div
              class="w-80 truncate"
              id="author"
            >
              <b>
                Authors:  
              </b>
              testAuthorName_1,testAuthorName_2 
            </div>
            <div
              id="durations"
            >
              <b>
                Duration:  
              </b>
              00:12 hours
            </div>
            <div
              id="created"
            >
              <b>
                Created:  
              </b>
              10.09.2022
            </div>
            <div
              id="showCourseBtn"
            >
              <div
                class="MuiBox-root css-0"
              >
                <div
                  class="MuiGrid-root MuiGrid-container mt-5 justify-items-center items-center css-11lq3yg-MuiGrid-root"
                >
                  <div
                    class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
                  >
                    <a
                      href="/testID"
                    >
                      <button
                        class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedSecondary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-root MuiButton-outlined MuiButton-outlinedSecondary MuiButton-sizeMedium MuiButton-outlinedSizeMedium css-wnpvqa-MuiButtonBase-root-MuiButton-root"
                        tabindex="0"
                        type="button"
                      >
                        Show Courses
                        <span
                          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                        />
                      </button>
                    </a>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
